bitwarden/ios
# [BITAU-153] [BITAU-144] [BITAU-208] Enable Background Syncing When the Phone Is Locked

## 🎟️ Tracking

[BITAU-144](https://livefront.atlassian.net/browse/BITAU-144)
[BITAU-153](https://livefront.atlassian.net/browse/BITAU-153)

## 📔 Objective

There is currently an error in the way we receive Push Notifications when the device is completely locked. All of the keys in the app are currently stored with the protection level `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`. This level of protection means that if we get a Push when the device is locked, none of the keys will be accessible. The first thing the NotificationService does is check `stateService.isAuthenticated()` which immediately tries to access the accessToken. This throws a security error, which is caught, logged, and fails silently. Therefore all background Push notifications will fail if the device is locked.

In addition, when we setup the Authenticator vault key to facilitate syncing to the Authenticator app, it was stored like all other keys with `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`. This was a known issue that would break syncing in the background if the device is locked.

This PR attempts to fix both of these situations by allowing some keys to be accessed in the background even if the device is locked:
- It refactors the way we setup permissions to allow for optionally specifying the protection level associated with the `KeychainItem`.
- This allows us to specify `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly` for the `accessToken` (which fixes Push) and the `authenticatorVaultKey` (which enables full background syncing to Authenticator).

With these changes in place, users can receive background pushes when the device is locked and the app will be able to process them. And those pushes can trigger the AuthenticatorSyncService to use it's key to unlock/lock the vault and perform the sync.

### ⚠️ Important Considerations and things to test ⚠️

#### Existing users will already have accessToken with the old permissions:
- When this rolls out in the app, all users will have previously stored `accessToken`s with the old protection level.
- `setAccessToken` like all keychain operations will first _delete_ the existing item before _adding_ the new value.
- From my testing so far, the delete succeeds, allowing us to write the new `accessToken` with the new protection level. **But this should be thoroughly tested**.
- This also means that a user will have to have written a new accessToken before they can have push and sync working with the device locked. Until they get the new protection level set, it will behave as it does now and fail silently.
 
#### Testing Steps
Here's a quick rundown of what I did to test this case:
- Ran the app with the old setting of device unlocked to make sure my account was setup. Logged in, etc.
- Locked vault and device and sent a push. Fails due to the `accessToken`'s protection level
- Rebuild to device with the NEW accessToken protection.
- Locked vault and device and sent another push. Fails again (same error, due to the fact that the `accessToken` hasn't been re-written and therefore has the old protection level)
- Unlocked everything, breakpointed the call for pending logins and sent back a `401`. This causes the app to fetch and store a new `accessToken`
- Verified that the setAccessToken call was going through and didn't have any errors. This means it should now have the new protection level.
- Locked vault and Device, sent push, and everything completes. Saw the notification go through, saw the sync service unlock vault, sync keys, lock it back up, etc.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines



- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes
